Ref #1589: Add warning messages for migration annotation

vitorsvieira opened the pull request:
Regarding #1589, this PR addresses:

- "{symbol} has an unparsable version number..." (RefCheck:694)
- "{symbol} has changed semantics in version..." (RefCheck:698)

nicolasstucki commented:
Please change the commit message to `Fix #1593: Add warning messages for migration annotation`

nicolasstucki commented on compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala:
```diff
@@ -1973,6 +1974,34 @@ object messages {
+    val explanation =
+      hl"""$migrationMessage
+          |
+          |The symbol is marked with ${"@migration"} indicating it has changed semantics
```
You should not talk about symbols as this is a compiler internal concept, the user should not know about them. You should probably replace it with `${symbol.showLocated}` or `${symbol.show}`.

nicolasstucki commented on compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala:
```diff
@@ -1973,6 +1974,34 @@ object messages {
+          |
+          |The symbol is marked with ${"@migration"} indicating it has changed semantics
+          |between versions and the ${"-Xmigration"} settings is used to warn about constructs
+          |whose behavior may have changed since version.""".stripMargin
```
"may have changed since version" -> "may have changed since version change"

nicolasstucki commented on compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala:
```diff
@@ -1973,6 +1974,34 @@ object messages {
+    val explanation = {
+      hl"""The symbol is marked with ${"@migration"} indicating it has changed semantics
+          |between versions and the ${"-Xmigration"} settings is used to warn about constructs
+          |whose behavior may have changed since version.""".stripMargin
```
same

nicolasstucki commented on compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala:
```diff
@@ -1973,6 +1974,34 @@ object messages {
+      hl"""${symbol.showLocated} has changed semantics in version $symbolVersion:
+          |$migrationVersion""".stripMargin
+    val explanation = {
+      hl"""The symbol is marked with ${"@migration"} indicating it has changed semantics
```
same (symbol)

vitorsvieira commented:
I was always curious about this annotation. Is there any reason why it is not part of the standard library? ATM, it is available only internally ` private[scala] final class migration(message: String, changedIn: String) extends scala.annotation.StaticAnnotation`.

Just out of curiosity, Swift provides the `@available` attribute, that IMO is an expanded 
 representation of scala's `@migration`. Maybe a combination of `@deprecated` and  `@migration` 

In Swift, `@available(Platform+VersionNumber*, Args*)` attribute can be used to indicate the whole `declaration`'s  lifecycle, relative to certain Swift language versions or certain platforms and operating system versions. Some of its arguments:

- unavailable(`version number`)
- introduced(`version number`)
- deprecated(`version number`)
- obsoleted(`version number`)
- message(`message`)
- renamed(`new name`)

In scala's ecosystem, this would mean the possibility to achieve something like:
`@available(JVM1_8, ScalaJS1_0, ScalaNative1_0, introduced:2.4, deprecated:2.6, obsoleted:2.10, renamed: "newMethodName")`

Reference: https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/Attributes.html#//apple_ref/doc/uid/TP40014097-CH35-ID347

PS: Not a suggestion, just toughts ;)

allanrenucci commented on compiler/src/dotty/tools/dotc/typer/RefChecks.scala:
```diff
@@ -691,11 +691,11 @@ object RefChecks {
       }
       if (changed)
-        ctx.warning(s"${sym.showLocated} has changed semantics in version $symVersion:\n${sym.migrationMessage.get}")
+        ctx.warning(SymbolChangedSemanticsInVersion(sym, symVersion, sym.migrationMessage.get))
```
While you're at it. Can you move this into the match. There is no need for this `changed` variable.
Also, `symVersion` is `sym.migrationVersion.get` so you don't need both of them

vitorsvieira replied:
Sure. Updated.

allanrenucci commented on compiler/src/dotty/tools/dotc/typer/RefChecks.scala:
```diff
@@ -685,17 +685,14 @@ object RefChecks {
-          ctx.settings.Xmigration.value < v
+    val xMigrationValue = ctx.settings.Xmigration.value
+    if (sym.hasAnnotation(defn.MigrationAnnot) && xMigrationValue != NoScalaVersion) {
+      sym.migrationVersion.get match {
```
Your match is not exhaustive

vitorsvieira replied:
Fixed.

allanrenucci commented on compiler/src/dotty/tools/dotc/typer/RefChecks.scala:
```diff
@@ -685,17 +685,14 @@ object RefChecks {
+    if (sym.hasAnnotation(defn.MigrationAnnot) && xMigrationValue != NoScalaVersion) {
+      sym.migrationVersion.get match {
+        case scala.util.Success(symVersion) if xMigrationValue < symVersion =>
+          ctx.warning(SymbolChangedSemanticsInVersion(sym, symVersion, sym.migrationMessage.get))
```
You can remove `sym.migrationMessage.get`. It is the same as `symVersion`. I don't know why it was duplicated in the first place

vitorsvieira replied:
I would have to change the final message if that is ok. Currently:
```
${symbol.showLocated} has changed semantics in version $symbolVersion:
$migrationVersion
```
The migrationMessage provides either a version string or an error message from `ScalaVersion.parse`:

```
 def failure = Failure(new NumberFormatException(
      s"There was a problem parsing ${versionString}. " +
       "Versions should be in the form major[.minor[.revision]] " +
       "where each part is a positive number, as in 2.10.1. " +
       "The minor and revision parts are optional."
    ))
```

allanrenucci replied:
It cannot be a failure since you're in the `Success` case of the match

vitorsvieira replied:
OK

allanrenucci commented on compiler/src/dotty/tools/dotc/typer/RefChecks.scala:
```diff
@@ -685,17 +685,14 @@ object RefChecks {
+    if (sym.hasAnnotation(defn.MigrationAnnot) && xMigrationValue != NoScalaVersion) {
+      sym.migrationVersion.get match {
+        case scala.util.Success(symVersion) if xMigrationValue < symVersion =>
+          ctx.warning(SymbolChangedSemanticsInVersion(sym, symVersion, sym.migrationMessage.get))
```
You're missing the `pos`

vitorsvieira replied:
Fixed.

allanrenucci commented on compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala:
```diff
@@ -1973,6 +1974,34 @@ object messages {
+          |
+          |The ${symbol.showLocated} is marked with ${"@migration"} indicating it has changed semantics
+          |between versions and the ${"-Xmigration"} settings is used to warn about constructs
+          |whose behavior may have changed since version change.""".stripMargin
```
You don't need to `stripMargin`

vitorsvieira replied:
Fixed.

allanrenucci commented on compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala:
```diff
@@ -1973,6 +1974,34 @@ object messages {
+    val explanation = {
+      hl"""The ${symbol.showLocated} is marked with ${"@migration"} indicating it has changed semantics
+          |between versions and the ${"-Xmigration"} settings is used to warn about constructs
+          |whose behavior may have changed since version change.""".stripMargin
```
You don't need to `stripMargin`

vitorsvieira replied:
Fixed.

allanrenucci commented on compiler/src/dotty/tools/dotc/typer/RefChecks.scala:
```diff
@@ -685,17 +685,15 @@ object RefChecks {
-          ctx.warning(s"${sym.showLocated} has an unparsable version number: ${ex.getMessage()}", pos)
-          false
+          ctx.warning(SymbolHasUnparsableVersionNumber(sym, ex.getMessage()), pos)
+        case _ => ()
```
Remove `()`

vitorsvieira replied:
Removed.
